[
   {
      "_links" : {
         "html" : {
            "href" : "https://github.com/bitcoin/bitcoin/pull/19532#discussion_r455935608"
         },
         "pull_request" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/19532"
         },
         "self" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/455935608"
         }
      },
      "author_association" : "CONTRIBUTOR",
      "body" : "This line appears to be getting caught in the linter. \r\n\r\n`test/functional/test_framework/test_framework.py:717:13: F401 'zmq' imported but unused`",
      "commit_id" : "e7f433e1e1223a2db05f195d32ef2c0e1537cd9b",
      "created_at" : "2020-07-16T17:00:58Z",
      "diff_hunk" : "@@ -714,7 +714,7 @@ def _initialize_chain_clean(self):\n     def skip_if_no_py3_zmq(self):\n         \"\"\"Attempt to import the zmq package and skip the test if the import fails.\"\"\"\n         try:\n-            import zmq  # noqa\n+            import zmq  # type:ignore",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/19532#discussion_r455935608",
      "id" : 455935608,
      "line" : 717,
      "node_id" : "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ1NTkzNTYwOA==",
      "original_commit_id" : "92b8001022c96eb70c2b10bfe8f459fae8a062c4",
      "original_line" : 717,
      "original_position" : 5,
      "original_start_line" : null,
      "path" : "test/functional/test_framework/test_framework.py",
      "position" : 5,
      "pull_request_review_id" : 450041576,
      "pull_request_url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/19532",
      "side" : "RIGHT",
      "start_line" : null,
      "start_side" : null,
      "updated_at" : "2020-07-16T17:00:58Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/455935608",
      "user" : {
         "avatar_url" : "https://avatars1.githubusercontent.com/u/755825?v=4",
         "events_url" : "https://api.github.com/users/adamjonas/events{/privacy}",
         "followers_url" : "https://api.github.com/users/adamjonas/followers",
         "following_url" : "https://api.github.com/users/adamjonas/following{/other_user}",
         "gists_url" : "https://api.github.com/users/adamjonas/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/adamjonas",
         "id" : 755825,
         "login" : "adamjonas",
         "node_id" : "MDQ6VXNlcjc1NTgyNQ==",
         "organizations_url" : "https://api.github.com/users/adamjonas/orgs",
         "received_events_url" : "https://api.github.com/users/adamjonas/received_events",
         "repos_url" : "https://api.github.com/users/adamjonas/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/adamjonas/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/adamjonas/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/adamjonas"
      }
   },
   {
      "_links" : {
         "html" : {
            "href" : "https://github.com/bitcoin/bitcoin/pull/19532#discussion_r455973891"
         },
         "pull_request" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/19532"
         },
         "self" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/455973891"
         }
      },
      "author_association" : "CONTRIBUTOR",
      "body" : "As far as I can tell, we don't be able to fix this short of disabling flake8 linting for the entire file.  It's a false positive, in this case the very act of importing the module is \"using\" it.",
      "commit_id" : "e7f433e1e1223a2db05f195d32ef2c0e1537cd9b",
      "created_at" : "2020-07-16T18:04:22Z",
      "diff_hunk" : "@@ -714,7 +714,7 @@ def _initialize_chain_clean(self):\n     def skip_if_no_py3_zmq(self):\n         \"\"\"Attempt to import the zmq package and skip the test if the import fails.\"\"\"\n         try:\n-            import zmq  # noqa\n+            import zmq  # type:ignore",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/19532#discussion_r455973891",
      "id" : 455973891,
      "in_reply_to_id" : 455935608,
      "line" : 717,
      "node_id" : "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ1NTk3Mzg5MQ==",
      "original_commit_id" : "92b8001022c96eb70c2b10bfe8f459fae8a062c4",
      "original_line" : 717,
      "original_position" : 5,
      "original_start_line" : null,
      "path" : "test/functional/test_framework/test_framework.py",
      "position" : 5,
      "pull_request_review_id" : 450089771,
      "pull_request_url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/19532",
      "side" : "RIGHT",
      "start_line" : null,
      "start_side" : null,
      "updated_at" : "2020-07-16T18:04:22Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/455973891",
      "user" : {
         "avatar_url" : "https://avatars2.githubusercontent.com/u/5553787?v=4",
         "events_url" : "https://api.github.com/users/troygiorshev/events{/privacy}",
         "followers_url" : "https://api.github.com/users/troygiorshev/followers",
         "following_url" : "https://api.github.com/users/troygiorshev/following{/other_user}",
         "gists_url" : "https://api.github.com/users/troygiorshev/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/troygiorshev",
         "id" : 5553787,
         "login" : "troygiorshev",
         "node_id" : "MDQ6VXNlcjU1NTM3ODc=",
         "organizations_url" : "https://api.github.com/users/troygiorshev/orgs",
         "received_events_url" : "https://api.github.com/users/troygiorshev/received_events",
         "repos_url" : "https://api.github.com/users/troygiorshev/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/troygiorshev/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/troygiorshev/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/troygiorshev"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "This all looks great!\r\n\r\nHowever, try as I might, I can't get MyPy to ever actually tell me something is wrong.  (i.e. I intentionally give a function an argument with the wrong type and MyPy still passes every file)  Have you been able to?\r\n\r\nRegarding your concern [here](https://github.com/bitcoin/bitcoin/issues/19389#issuecomment-659054074) that this isn't python 3.5 compatible, I don't think you have to worry.  The only parts that aren't compatible are the `.pyi` files, but that's ok, they're only used by MyPy.  I've tested running `lint-python.sh` under python 3.5 and it doesn't give any errors.  (MyPy does, however, get confused and tell me that a couple lines are wrong when they aren't.  It doesn't seem related though).",
      "created_at" : "2020-07-16T18:14:55Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/19532#issuecomment-659584317",
      "id" : 659584317,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/19532",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDY1OTU4NDMxNw==",
      "updated_at" : "2020-07-16T18:24:08Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/659584317",
      "user" : {
         "avatar_url" : "https://avatars2.githubusercontent.com/u/5553787?v=4",
         "events_url" : "https://api.github.com/users/troygiorshev/events{/privacy}",
         "followers_url" : "https://api.github.com/users/troygiorshev/followers",
         "following_url" : "https://api.github.com/users/troygiorshev/following{/other_user}",
         "gists_url" : "https://api.github.com/users/troygiorshev/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/troygiorshev",
         "id" : 5553787,
         "login" : "troygiorshev",
         "node_id" : "MDQ6VXNlcjU1NTM3ODc=",
         "organizations_url" : "https://api.github.com/users/troygiorshev/orgs",
         "received_events_url" : "https://api.github.com/users/troygiorshev/received_events",
         "repos_url" : "https://api.github.com/users/troygiorshev/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/troygiorshev/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/troygiorshev/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/troygiorshev"
      }
   },
   {
      "_links" : {
         "html" : {
            "href" : "https://github.com/bitcoin/bitcoin/pull/19532#discussion_r456005243"
         },
         "pull_request" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/19532"
         },
         "self" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/456005243"
         }
      },
      "author_association" : "CONTRIBUTOR",
      "body" : "It's worth noting that this brings us from 3 `F401` warnings up to 4 `F401` warnings.",
      "commit_id" : "e7f433e1e1223a2db05f195d32ef2c0e1537cd9b",
      "created_at" : "2020-07-16T18:58:49Z",
      "diff_hunk" : "@@ -714,7 +714,7 @@ def _initialize_chain_clean(self):\n     def skip_if_no_py3_zmq(self):\n         \"\"\"Attempt to import the zmq package and skip the test if the import fails.\"\"\"\n         try:\n-            import zmq  # noqa\n+            import zmq  # type:ignore",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/19532#discussion_r456005243",
      "id" : 456005243,
      "in_reply_to_id" : 455935608,
      "line" : 717,
      "node_id" : "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ1NjAwNTI0Mw==",
      "original_commit_id" : "92b8001022c96eb70c2b10bfe8f459fae8a062c4",
      "original_line" : 717,
      "original_position" : 5,
      "original_start_line" : null,
      "path" : "test/functional/test_framework/test_framework.py",
      "position" : 5,
      "pull_request_review_id" : 450128178,
      "pull_request_url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/19532",
      "side" : "RIGHT",
      "start_line" : null,
      "start_side" : null,
      "updated_at" : "2020-07-16T18:58:50Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/456005243",
      "user" : {
         "avatar_url" : "https://avatars2.githubusercontent.com/u/5553787?v=4",
         "events_url" : "https://api.github.com/users/troygiorshev/events{/privacy}",
         "followers_url" : "https://api.github.com/users/troygiorshev/followers",
         "following_url" : "https://api.github.com/users/troygiorshev/following{/other_user}",
         "gists_url" : "https://api.github.com/users/troygiorshev/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/troygiorshev",
         "id" : 5553787,
         "login" : "troygiorshev",
         "node_id" : "MDQ6VXNlcjU1NTM3ODc=",
         "organizations_url" : "https://api.github.com/users/troygiorshev/orgs",
         "received_events_url" : "https://api.github.com/users/troygiorshev/received_events",
         "repos_url" : "https://api.github.com/users/troygiorshev/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/troygiorshev/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/troygiorshev/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/troygiorshev"
      }
   },
   {
      "_links" : {
         "html" : {
            "href" : "https://github.com/bitcoin/bitcoin/pull/19532#discussion_r456293635"
         },
         "pull_request" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/19532"
         },
         "self" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/456293635"
         }
      },
      "author_association" : "MEMBER",
      "body" : "is it not possible to keep the `# noqa` linter disable comment?",
      "commit_id" : "e7f433e1e1223a2db05f195d32ef2c0e1537cd9b",
      "created_at" : "2020-07-17T08:20:36Z",
      "diff_hunk" : "@@ -714,7 +714,7 @@ def _initialize_chain_clean(self):\n     def skip_if_no_py3_zmq(self):\n         \"\"\"Attempt to import the zmq package and skip the test if the import fails.\"\"\"\n         try:\n-            import zmq  # noqa\n+            import zmq  # type:ignore",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/19532#discussion_r456293635",
      "id" : 456293635,
      "in_reply_to_id" : 455935608,
      "line" : 717,
      "node_id" : "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ1NjI5MzYzNQ==",
      "original_commit_id" : "92b8001022c96eb70c2b10bfe8f459fae8a062c4",
      "original_line" : 717,
      "original_position" : 5,
      "original_start_line" : null,
      "path" : "test/functional/test_framework/test_framework.py",
      "position" : 5,
      "pull_request_review_id" : 450464517,
      "pull_request_url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/19532",
      "side" : "RIGHT",
      "start_line" : null,
      "start_side" : null,
      "updated_at" : "2020-07-17T08:20:37Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/456293635",
      "user" : {
         "avatar_url" : "https://avatars3.githubusercontent.com/u/1063656?v=4",
         "events_url" : "https://api.github.com/users/jnewbery/events{/privacy}",
         "followers_url" : "https://api.github.com/users/jnewbery/followers",
         "following_url" : "https://api.github.com/users/jnewbery/following{/other_user}",
         "gists_url" : "https://api.github.com/users/jnewbery/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/jnewbery",
         "id" : 1063656,
         "login" : "jnewbery",
         "node_id" : "MDQ6VXNlcjEwNjM2NTY=",
         "organizations_url" : "https://api.github.com/users/jnewbery/orgs",
         "received_events_url" : "https://api.github.com/users/jnewbery/received_events",
         "repos_url" : "https://api.github.com/users/jnewbery/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/jnewbery/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/jnewbery/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/jnewbery"
      }
   },
   {
      "_links" : {
         "html" : {
            "href" : "https://github.com/bitcoin/bitcoin/pull/19532#discussion_r456408788"
         },
         "pull_request" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/19532"
         },
         "self" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/456408788"
         }
      },
      "author_association" : "CONTRIBUTOR",
      "body" : "Unfortunately not.  We need the `type:ignore` or else MyPy fails completely.  I've tried things like putting them both on the same line but it looks like flake8 and MyPy just don't play nicely together in this sort of conflict.\r\n\r\nWorth noting maybe that pylint allows us to disable a particular linting error for an entire function (which would fix this problem), where AFAICT flake8 does not.",
      "commit_id" : "e7f433e1e1223a2db05f195d32ef2c0e1537cd9b",
      "created_at" : "2020-07-17T12:25:39Z",
      "diff_hunk" : "@@ -714,7 +714,7 @@ def _initialize_chain_clean(self):\n     def skip_if_no_py3_zmq(self):\n         \"\"\"Attempt to import the zmq package and skip the test if the import fails.\"\"\"\n         try:\n-            import zmq  # noqa\n+            import zmq  # type:ignore",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/19532#discussion_r456408788",
      "id" : 456408788,
      "in_reply_to_id" : 455935608,
      "line" : 717,
      "node_id" : "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ1NjQwODc4OA==",
      "original_commit_id" : "92b8001022c96eb70c2b10bfe8f459fae8a062c4",
      "original_line" : 717,
      "original_position" : 5,
      "original_start_line" : null,
      "path" : "test/functional/test_framework/test_framework.py",
      "position" : 5,
      "pull_request_review_id" : 450612713,
      "pull_request_url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/19532",
      "side" : "RIGHT",
      "start_line" : null,
      "start_side" : null,
      "updated_at" : "2020-07-17T12:25:57Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/456408788",
      "user" : {
         "avatar_url" : "https://avatars2.githubusercontent.com/u/5553787?v=4",
         "events_url" : "https://api.github.com/users/troygiorshev/events{/privacy}",
         "followers_url" : "https://api.github.com/users/troygiorshev/followers",
         "following_url" : "https://api.github.com/users/troygiorshev/following{/other_user}",
         "gists_url" : "https://api.github.com/users/troygiorshev/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/troygiorshev",
         "id" : 5553787,
         "login" : "troygiorshev",
         "node_id" : "MDQ6VXNlcjU1NTM3ODc=",
         "organizations_url" : "https://api.github.com/users/troygiorshev/orgs",
         "received_events_url" : "https://api.github.com/users/troygiorshev/received_events",
         "repos_url" : "https://api.github.com/users/troygiorshev/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/troygiorshev/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/troygiorshev/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/troygiorshev"
      }
   },
   {
      "_links" : {
         "html" : {
            "href" : "https://github.com/bitcoin/bitcoin/pull/19532#discussion_r456421041"
         },
         "pull_request" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/19532"
         },
         "self" : {
            "href" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/456421041"
         }
      },
      "author_association" : "MEMBER",
      "body" : "that's unfortunate :(",
      "commit_id" : "e7f433e1e1223a2db05f195d32ef2c0e1537cd9b",
      "created_at" : "2020-07-17T12:50:02Z",
      "diff_hunk" : "@@ -714,7 +714,7 @@ def _initialize_chain_clean(self):\n     def skip_if_no_py3_zmq(self):\n         \"\"\"Attempt to import the zmq package and skip the test if the import fails.\"\"\"\n         try:\n-            import zmq  # noqa\n+            import zmq  # type:ignore",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/19532#discussion_r456421041",
      "id" : 456421041,
      "in_reply_to_id" : 455935608,
      "line" : 717,
      "node_id" : "MDI0OlB1bGxSZXF1ZXN0UmV2aWV3Q29tbWVudDQ1NjQyMTA0MQ==",
      "original_commit_id" : "92b8001022c96eb70c2b10bfe8f459fae8a062c4",
      "original_line" : 717,
      "original_position" : 5,
      "original_start_line" : null,
      "path" : "test/functional/test_framework/test_framework.py",
      "position" : 5,
      "pull_request_review_id" : 450628264,
      "pull_request_url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/19532",
      "side" : "RIGHT",
      "start_line" : null,
      "start_side" : null,
      "updated_at" : "2020-07-17T12:50:02Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/pulls/comments/456421041",
      "user" : {
         "avatar_url" : "https://avatars3.githubusercontent.com/u/1063656?v=4",
         "events_url" : "https://api.github.com/users/jnewbery/events{/privacy}",
         "followers_url" : "https://api.github.com/users/jnewbery/followers",
         "following_url" : "https://api.github.com/users/jnewbery/following{/other_user}",
         "gists_url" : "https://api.github.com/users/jnewbery/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/jnewbery",
         "id" : 1063656,
         "login" : "jnewbery",
         "node_id" : "MDQ6VXNlcjEwNjM2NTY=",
         "organizations_url" : "https://api.github.com/users/jnewbery/orgs",
         "received_events_url" : "https://api.github.com/users/jnewbery/received_events",
         "repos_url" : "https://api.github.com/users/jnewbery/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/jnewbery/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/jnewbery/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/jnewbery"
      }
   },
   {
      "author_association" : "NONE",
      "body" : "> This all looks great!\r\n> \r\n> However, try as I might, I can't get MyPy to ever actually tell me something is wrong. (i.e. I intentionally give a function an argument with the wrong type and MyPy still passes every file) Have you been able to?\r\n> \r\n> Regarding your concern [here](https://github.com/bitcoin/bitcoin/issues/19389#issuecomment-659054074) that this isn't python 3.5 compatible, I don't think you have to worry. The only parts that aren't compatible are the `.pyi` files, but that's ok, they're only used by MyPy. I've tested running `lint-python.sh` under python 3.5 and it doesn't give any errors. (MyPy does, however, get confused and tell me that a couple lines are wrong when they aren't. It doesn't seem related though).\r\n\r\nThanks for taking the time to review the code! \r\n\r\nWhen I run the linter file (lint-python.py), Mypy doesn't catch any errors, however, when I run mypy from the command line as such:\r\n`MYPYPATH=/path/to/stub/files mypy test/functional/contains_mypy_errors.py` \r\nincompatible type errors are properly caught and displayed. I haven't been able to find too much on this unexpected behavior from Mypy other than this Stackoverflow post: https://stackoverflow.com/questions/55437746/how-do-i-correctly-set-mypypath-to-pick-up-stubs-for-mypy.",
      "created_at" : "2020-07-18T16:59:23Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/19532#issuecomment-660510955",
      "id" : 660510955,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/19532",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDY2MDUxMDk1NQ==",
      "performed_via_github_app" : null,
      "updated_at" : "2020-07-18T16:59:23Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/660510955",
      "user" : {
         "avatar_url" : "https://avatars3.githubusercontent.com/u/24584383?v=4",
         "events_url" : "https://api.github.com/users/Ghorbanian/events{/privacy}",
         "followers_url" : "https://api.github.com/users/Ghorbanian/followers",
         "following_url" : "https://api.github.com/users/Ghorbanian/following{/other_user}",
         "gists_url" : "https://api.github.com/users/Ghorbanian/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/Ghorbanian",
         "id" : 24584383,
         "login" : "Ghorbanian",
         "node_id" : "MDQ6VXNlcjI0NTg0Mzgz",
         "organizations_url" : "https://api.github.com/users/Ghorbanian/orgs",
         "received_events_url" : "https://api.github.com/users/Ghorbanian/received_events",
         "repos_url" : "https://api.github.com/users/Ghorbanian/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/Ghorbanian/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/Ghorbanian/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/Ghorbanian"
      }
   },
   {
      "author_association" : "NONE",
      "body" : "I'm new to mypy, but I found this [issue](https://github.com/python/mypy/issues/1645) which seems to address the problem that was previously encountered. It seems like the `--namespace-packages` flag will fix the issue, but it may be easier to just add the empty `__init__.py` file to `test/functional/data` which fixes the problem too.\r\n\r\nTo test this,  I included the `__init__.py` file in the `data` directory and then `mypy test/functional` runs without any errors. Creating an obvious type error inside the `data` package also results in the error being correctly detected. This will make the generated stub files no longer necessary, so the type annotations can be made directly in the source files.",
      "created_at" : "2020-08-09T20:08:01Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/19532#issuecomment-671095289",
      "id" : 671095289,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/19532",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDY3MTA5NTI4OQ==",
      "performed_via_github_app" : null,
      "updated_at" : "2020-08-09T20:08:01Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/671095289",
      "user" : {
         "avatar_url" : "https://avatars3.githubusercontent.com/u/29065991?v=4",
         "events_url" : "https://api.github.com/users/adaminsky/events{/privacy}",
         "followers_url" : "https://api.github.com/users/adaminsky/followers",
         "following_url" : "https://api.github.com/users/adaminsky/following{/other_user}",
         "gists_url" : "https://api.github.com/users/adaminsky/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/adaminsky",
         "id" : 29065991,
         "login" : "adaminsky",
         "node_id" : "MDQ6VXNlcjI5MDY1OTkx",
         "organizations_url" : "https://api.github.com/users/adaminsky/orgs",
         "received_events_url" : "https://api.github.com/users/adaminsky/received_events",
         "repos_url" : "https://api.github.com/users/adaminsky/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/adaminsky/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/adaminsky/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/adaminsky"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "@adaminsky thanks for looking at this!  Great to see that we might not need the stub files after all!  I don't think we're all the way there yet thought, creating `test/functional/data/__init__.py` is already the second commit of this PR :)\r\n\r\nDid you try making a type error inside a test in `test/functional/`?  I still can't get those to be caught.",
      "created_at" : "2020-08-10T21:19:52Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/19532#issuecomment-671595547",
      "id" : 671595547,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/19532",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDY3MTU5NTU0Nw==",
      "performed_via_github_app" : null,
      "updated_at" : "2020-08-10T21:19:52Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/671595547",
      "user" : {
         "avatar_url" : "https://avatars2.githubusercontent.com/u/5553787?v=4",
         "events_url" : "https://api.github.com/users/troygiorshev/events{/privacy}",
         "followers_url" : "https://api.github.com/users/troygiorshev/followers",
         "following_url" : "https://api.github.com/users/troygiorshev/following{/other_user}",
         "gists_url" : "https://api.github.com/users/troygiorshev/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/troygiorshev",
         "id" : 5553787,
         "login" : "troygiorshev",
         "node_id" : "MDQ6VXNlcjU1NTM3ODc=",
         "organizations_url" : "https://api.github.com/users/troygiorshev/orgs",
         "received_events_url" : "https://api.github.com/users/troygiorshev/received_events",
         "repos_url" : "https://api.github.com/users/troygiorshev/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/troygiorshev/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/troygiorshev/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/troygiorshev"
      }
   },
   {
      "author_association" : "CONTRIBUTOR",
      "body" : "I've tried changing MyPy's configuration options, maybe that will do the trick.  To try this too, create a config file, say, `test/functional/data/mypy.ini` with the following in it:\r\n\r\n```\r\n[mypy]\r\ncheck_untyped_defs = True\r\nno_implicit_optional = True\r\n```\r\n\r\nThen running MyPy with `mypy --config-file test/functional/data/mypy.ini test/functional` gives tonnes of errors (finally ð) but I'm not sure that all of them are useful yet.",
      "created_at" : "2020-08-10T21:20:13Z",
      "html_url" : "https://github.com/bitcoin/bitcoin/pull/19532#issuecomment-671595692",
      "id" : 671595692,
      "issue_url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/19532",
      "node_id" : "MDEyOklzc3VlQ29tbWVudDY3MTU5NTY5Mg==",
      "performed_via_github_app" : null,
      "updated_at" : "2020-08-10T21:20:13Z",
      "url" : "https://api.github.com/repos/bitcoin/bitcoin/issues/comments/671595692",
      "user" : {
         "avatar_url" : "https://avatars2.githubusercontent.com/u/5553787?v=4",
         "events_url" : "https://api.github.com/users/troygiorshev/events{/privacy}",
         "followers_url" : "https://api.github.com/users/troygiorshev/followers",
         "following_url" : "https://api.github.com/users/troygiorshev/following{/other_user}",
         "gists_url" : "https://api.github.com/users/troygiorshev/gists{/gist_id}",
         "gravatar_id" : "",
         "html_url" : "https://github.com/troygiorshev",
         "id" : 5553787,
         "login" : "troygiorshev",
         "node_id" : "MDQ6VXNlcjU1NTM3ODc=",
         "organizations_url" : "https://api.github.com/users/troygiorshev/orgs",
         "received_events_url" : "https://api.github.com/users/troygiorshev/received_events",
         "repos_url" : "https://api.github.com/users/troygiorshev/repos",
         "site_admin" : false,
         "starred_url" : "https://api.github.com/users/troygiorshev/starred{/owner}{/repo}",
         "subscriptions_url" : "https://api.github.com/users/troygiorshev/subscriptions",
         "type" : "User",
         "url" : "https://api.github.com/users/troygiorshev"
      }
   }
]
